### PR TITLE
Links and images

### LINKED ISSUE #9: Strings that count as absolute URIs need not be valid URIs
See <https://spec.commonmark.org/0.30/#example-597>

### LINKED ISSUE #10: Spaces are not allowed in autolinks
See <https://spec.commonmark.org/0.30/#example-601>

### LINKED ISSUE #39: An image description may have inline elements as its contents
See <https://spec.commonmark.org/0.30/#example-572>

### LINKED ISSUE #47: The destination cannot contain line endings, even if enclosed in pointy brackets
See <https://spec.commonmark.org/0.30/#example-490>

### LINKED ISSUE #48: Pointy brackets that enclose links must be unescaped
See <https://spec.commonmark.org/0.30/#example-492>

### LINKED ISSUE #49: Nested balanced quotes are not allowed without escaping
See <https://spec.commonmark.org/0.30/#example-507>

### LINKED ISSUE #50: Spaces, tabs, and up to one line ending is allowed around the destination and title
See <https://spec.commonmark.org/0.30/#example-509>

### LINKED ISSUE #51: Spaces, tabs, nor any line endings are allowed between the link text and the following parenthesis
See <https://spec.commonmark.org/0.30/#example-510>

### LINKED ISSUE #52: The link text may contain balanced brackets, but not unbalanced ones, unless they are escaped
See <https://spec.commonmark.org/0.30/#example-514>

### LINKED ISSUE #53: Links may not contain other links, at any level of nesting
See <https://spec.commonmark.org/0.30/#example-519>

### LINKED ISSUE #54: Brackets that aren’t part of links do not take precedence
See <https://spec.commonmark.org/0.30/#example-522>

### LINKED ISSUE #57: No spaces, tabs, or line endings are allowed between the link text and the link label
See <https://spec.commonmark.org/0.30/#example-541>

### LINKED ISSUE #59: Link labels cannot contain brackets, unless they are backslash-escaped
See <https://spec.commonmark.org/0.30/#example-546> and <https://spec.commonmark.org/0.30/#example-547>

### LINKED ISSUE #60: A link label must contain at least one character that is not a space, tab, or line ending
See <https://spec.commonmark.org/0.30/#example-550>

### LINKED ISSUE #61: If a link text consists of multiple invalid link labels, the last is used, others are parsed as text
See <https://spec.commonmark.org/0.30/#example-568>

### LINKED ISSUE #62: A link label is not parsed as shortcut reference when followed by a link label without a reference
See <https://spec.commonmark.org/0.30/#example-570>

### LINKED ISSUE #103: Link labels can contain escaped backslashes
See <https://spec.commonmark.org/0.30/#example-549>

### PR DESCRIPTION
**Pull request by lostenderman**
Closes #9
Closes #10
Closes #47
Closes #48

Closes #39
Closes #49
Closes #50
Closes #51
Closes #52
Closes #53
Closes #54
Closes #57
Closes #59
Closes #60
Closes #61
Closes #62
Closes #103

### CODE REVIEW

File: markdown.dtx [hunk 1]
```diff
@@ -21289,6 +21287,38 @@ end
...
+
+parsers.autolink_scheme = parsers.letter * parsers.autolink_scheme_part
+                        * parsers.autolink_scheme_part^-30
+
+parsers.absolute_uri = parsers.autolink_scheme * parsers.colon 
+                      * (parsers.any - parsers.spacing - parsers.less - parsers.more)^0
+
+parsers.email_address_dns_label = parsers.alphanumeric * ((parsers.alphanumeric + parsers.dash)^-62 * B(parsers.alphanumeric))^-1
+
+parsers.email_address = (parsers.alphanumeric + S(".!#$%&'*+/=?^_`{|}~-"))^1
```
Comment (Witiko) [inline markdown.dtx:21302]: ``S(".!#$%&'*+/=?^_`{|}~-")`` seems to be a good candidate for getting a name.

File: markdown.dtx [hunk 2]
```diff
@@ -21317,32 +21347,45 @@ parsers.dquoted     = P{ parsers.dquote * parsers.alphanumeric
...
+parsers.link_label  = parsers.lbracket 
+                    * -#(parsers.sp * parsers.rbracket) 
+                    * #((parsers.any - parsers.rbracket)^-999 * parsers.rbracket)
+                    * Cs((parsers.alphanumeric^1
+                       + parsers.inticks
+                       + parsers.autolink
+                       + ( parsers.backslash * parsers.backslash)
+                       + ( parsers.backslash * (parsers.lbracket + parsers.rbracket)
+                         + parsers.any
+                         - (parsers.lbracket + parsers.rbracket + parsers.blankline^2)))^1)
```
Comment (Witiko) [inline markdown.dtx:21370]:
```suggestion
                        + parsers.inticks
                        + parsers.autolink
                        + ( parsers.backslash * parsers.backslash)
                        + ( parsers.backslash * (parsers.lbracket + parsers.rbracket)
                          + parsers.any
                          - (parsers.lbracket + parsers.rbracket + parsers.blankline^2)))^1)
```

File: markdown.dtx [hunk 3]
```diff
@@ -21317,32 +21347,45 @@ parsers.dquoted     = P{ parsers.dquote * parsers.alphanumeric
...
 
 -- url for markdown links, allowing nested brackets:
 parsers.url         = parsers.less * Cs((parsers.anyescaped
-                                        - parsers.more)^0)
+                                        - parsers.newline - parsers.more)^0)
                                    * parsers.more
-                    + Cs((parsers.inparens + (parsers.anyescaped
+                    + -parsers.less * Cs((parsers.inparens + (parsers.anyescaped
                                              - parsers.spacing
                                              - parsers.rparent))^1)
```
Comment (Witiko) [inline markdown.dtx:21379]:
```suggestion
                    + -parsers.less
                    * Cs((parsers.inparens + ( parsers.anyescaped
                                             - parsers.spacing
                                             - parsers.rparent))^1)
```

File: markdown.dtx [hunk 4]
```diff
@@ -21983,57 +22384,28 @@ function M.reader.new(writer, options)
...
-  parsers.DirectImage   = parsers.exclamation
-                        * (parsers.tag / self.parser_functions.parse_inlines)
-                        * parsers.spnl
-                        * parsers.lparent
-                        * (parsers.url + Cc(""))  -- link can be empty [foo]()
-                        * parsers.optionaltitle
-                        * parsers.rparent
-                        / writer.image
+  parsers.MaybeLinkOrImage = Ct(parsers.link_array)  
+                          / defer_link_and_image_processing
```
Comment (Witiko) [inline markdown.dtx:22405]:
```suggestion
                           / defer_link_and_image_processing
```

File: markdown.dtx
Comment (Witiko) [inline markdown.dtx:1]: As in #138, this is an extensive addition in a style quite foreign to the rest of the codebase. Can you please add a sentence or two of prose above each function, documenting what it does and what the purpose of the individual parameters is? We will need it soon enough when resolving merge conflicts between this pull request and https://github.com/Witiko/markdown/pull/256.

**Review by Witiko — CHANGES_REQUESTED** [review]

File: markdown.dtx [hunk 5]
```diff
@@ -21285,23 +21285,41 @@ end
...
+                              + parsers.dash
 
-parsers.autolink_scheme = parsers.letter * parsers.autolink_scheme_part
+parsers.autolink_scheme = parsers.letter
+                        * parsers.autolink_scheme_part
                         * parsers.autolink_scheme_part^-30
 
-parsers.absolute_uri = parsers.autolink_scheme * parsers.colon 
+parsers.absolute_uri = parsers.autolink_scheme * parsers.colon
                       * (parsers.any - parsers.spacing - parsers.less - parsers.more)^0
```
Comment (Witiko) [inline markdown.dtx]:
```suggestion
parsers.absolute_uri = parsers.autolink_scheme * parsers.colon
                     * (parsers.any - parsers.spacing - parsers.less - parsers.more)^0
```

File: markdown.dtx [hunk 6]
```diff
@@ -21956,31 +21975,33 @@ function M.reader.new(writer, options)
...
+                                  * Cg(C(((parsers.backslash^-1 / "" * parsers.any - parsers.link_image_open_or_close))^0), "content"))
 
   parsers.link_array = C((parsers.exclamation^-1 * parsers.any_link)^1)
 
   parsers.link_group = Ct( parsers.link_image_opening
                         * (parsers.link_image_content * parsers.link_image_open_or_close)^0)
 
+  -- return the dangling parentheses of an inline link
+  -- exactly as they were in the input
   local function resolve_inline_link_closer(t, index, include_bracket)
```
Comment (Witiko) [inline markdown.dtx]: This is a definite improvement over not having any explanation, although I would still appreciate full-blown prose that can be typeset in the technical documentation and that allows you to describe the function in detail. For example, the following is preferable (here and everywhere):

```suggestion
%    \end{macrocode}
% \begin{markdown}
%
% Return the dangling parentheses of an inline link exactly as they were in the input.
%
% \end{markdown}
%  \begin{macrocode}
```

Keeping the description grounded in the actual parameters of the function is also appreciated. For example, you talk about "input" in your description, but the function accepts `t`, `index`, and `include_brackets` as parameters. The desciption should describe the intent of the function but also help the reader to understand what the parameters are.

**Comment by lostenderman** [comment]
I suspect that the implementation of emphasis (and strong emphasis) will have to be merged with the implementation of links (and images) similarly to [this](https://spec.commonmark.org/0.30/#an-algorithm-for-parsing-nested-emphasis-and-links) essentially, because of these examples:

[522](https://spec.commonmark.org/0.30/#example-522):
```
*foo [bar* baz]
```

This is an emphasis because the link has no reference.

[533](https://spec.commonmark.org/0.30/#example-533):
```
*[foo*][ref]

[ref]: /uri
```

This is not an emphasis because the link has a reference.

Currently, only the second example would be parsed correctly. In the first one, no emphasis would be produced since the link is matched,  even though it does not produce a link.

File: markdown.dtx [hunk 7]
```diff
@@ -24159,10 +24198,10 @@ function M.reader.new(writer, options)
...
                      end, false)
 
-  self.create_parser("parse_inlines_no_link",
-                     function()
-                       return parsers.inlines_no_link
-                     end, false)
+  self.create_parser("parse_inlines_full_link_and_image",
+                  function()
+                    return parsers.inlines_full_link_and_image
+                  end, false)
```
Comment (Witiko) [inline markdown.dtx:24204]:
```suggestion
  self.create_parser("parse_inlines_full_link_and_image",
                     function()
                       return parsers.inlines_full_link_and_image
                     end, false)
```

File: markdown.dtx [hunk 8]
```diff
@@ -24244,119 +24283,661 @@ function M.reader.new(writer, options)
...
+
+%    \end{macrocode}
+% \begin{markdown}
+%
+% Return the parentheses of an inline link for a closing delimiter at index `index` in the delimiter table `t`.
+% If the \Opt{linkAttributes} option extension is enabled, then also the attributes are returned.
+% The closing bracket is returned when the `include_bracket` is selected.
+%
+% \end{markdown}
+%  \begin{macrocode}
```
Comment (Witiko) [inline markdown.dtx:24422]: Same comment as https://github.com/lostenderman/markdown/pull/141/files#r1154487062.

File: markdown.dtx [hunk 9]
```diff
@@ -26453,93 +27025,49 @@ M.extensions.link_attributes = function()
...
-      local IndirectImageWithAttributes = parsers.indirect_image
-                                        * C(Ct(parsers.attributes))
-                                        / indirect_image
+      local MaybeLinkOrImageWithAttributes  = Ct(link_array) * Cc(link_group)
+                                              / defer_link_and_image_processing
 
-      local ImageWithAttributes = DirectImageWithAttributes
-                                + IndirectImageWithAttributes
+      local LinkOrImageWithAttributes     = Ct(link_array) * Cc(link_group)
+                                            / process_links_and_images
```
Comment (Witiko) [inline markdown.dtx:27067]:
```suggestion
      local LinkOrImageWithAttributes       = Ct(link_array) * Cc(link_group)
                                            / process_links_and_images
```

**Review by Witiko — CHANGES_REQUESTED** [review]
I fixed some style issues and commented on a limitation that I am willing to merge but should be highlighted in the thesis. Overall, I am somewhat blind without CI, so I will leave more in-depth reading until https://github.com/Witiko/markdown/pull/226. Feel free to merge this (or any other) PR.

**Pull request merged**